gsd/gsd.c

```diff
@@ -749,51 +679,872 @@ int gsd_read_header(struct gsd_handle* handle)
+    @param end Last index of the valid hep in *buf*.
+*/
+inline static void gsd_heap_shift_down(struct gsd_index_buffer* buf, size_t start, size_t end)
+{
+    size_t root = start;
+
+    while (gsd_heap_left_child(root) <= end)
+    {
+        size_t child = gsd_heap_left_child(root);
+        size_t swap = root;
+
+        if (gsd_cmp_index_entry(buf->data + swap, buf->data + child) < 0)
+        {
+            swap = child;
+        }
+        if (child + 1 <= end && gsd_cmp_index_entry(buf->data + swap, buf->data + child + 1) < 0)
+        {
+            swap = child + 1;
+        }
+
+        if (swap == root)
+        {
+            return;
+        }
+
+        gsd_heap_swap(buf, root, swap);
+        root = swap;
+    }
+}
+
+/** @internal
+    @brief Convert unordered index buffer to a heap
+    @param buf Buffer.
+*/
+inline static void gsd_heapify(struct gsd_index_buffer* buf)
+{
+    ssize_t start = gsd_heap_parent(buf->size - 1);
+
+    while (start >= 0)
+    {
+        gsd_heap_shift_down(buf, start, buf->size - 1);
+        start--;
+    }
+}
+
+/** @internal
+    @brief Sort the index buffer.
+
+    @param buf Buffer to sort.
+
+    Sorts an in-memory index buffer. Does not accept mapped indices.
+
+    @returns GSD_SUCCESS on success, GSD_* error codes on error.
+*/
+inline static int gsd_index_buffer_sort(struct gsd_index_buffer* buf)
+{
+    if (buf == NULL || buf->mapped_data || buf->reserved == 0)
+    {
+        return GSD_ERROR_INVALID_ARGUMENT;
+    }
+
+    // arrays of size 0 or 1 are already sorted
+    if (buf->size <= 1)
+    {
+        return GSD_SUCCESS;
+    }
+
+    gsd_heapify(buf);
+
+    size_t end = buf->size - 1;
+    while (end > 0)
+    {
+        gsd_heap_swap(buf, end, 0);
+        end = end - 1;
+        gsd_heap_shift_down(buf, 0, end);
+    }
+
+    return GSD_SUCCESS;
+}
+
+/** @internal
+    @brief Utility function to expand the memory space for the index block in the file.
+
+    @param handle Handle to the open gsd file.
+    @param size_required The new index must be able to hold at least this many elements.
+
+    @returns GSD_SUCCESS on success, GSD_* error codes on error.
+*/
+static int gsd_expand_file_index(struct gsd_handle* handle, size_t size_required)
+{
+    if (handle->open_flags == GSD_OPEN_READONLY)
+    {
+        return GSD_ERROR_FILE_MUST_BE_WRITABLE;
+    }
+
+    // multiply the index size each time it grows
+    // this allows the index to grow rapidly to accommodate new frames
+    const int multiplication_factor = 2;
+
+    // save the old size and update the new size
+    size_t size_old = handle->header.index_allocated_entries;
+    size_t size_new = size_old * multiplication_factor;
+
+    while (size_new <= size_required)
+    {
+        size_new *= multiplication_factor;
+    }
+
+    // Mac systems deadlock when writing from a mapped region into the tail end of that same region
+    // unmap the index first and copy it over by chunks
```

Comment:
Is it worth `#ifdef`ing to directly copy on non-Mac systems?

Reply:
I don't think so. This method is called rarely and only having 1 implementation to test and debug is a benefit. Changing from the memory mapped read to copying by chunks didn't change performance in any of my measurements.